### DISCUSSION TITLE
How exactly are the algorithms in here allowed/intended to be used?

### DISCUSSION THREAD

**Discussion by MultisampledNight** [Myndex/SAPC-APCA#66] [Q&A] [answered]
Currently I'm building a Rust library for doing calculations with color, as the current options don't quite fit for what I'd want. Searching for ways of calculating text contrast, I found

- https://www.w3.org/WAI/GL/wiki/Relative_luminance which linked to
- https://www.w3.org/WAI/GL/task-forces/silver/wiki/Visual_Contrast_of_Text_Subgroup which linked to
- https://www.w3.org/WAI/GL/task-forces/silver/wiki/Visual_Contrast_of_Text_Subgroup/APCA_model which finally linked to this repository.

However, while looking through the source, I didn't find any kind of license, only this line at various places:

https://github.com/Myndex/SAPC-APCA/blob/dbc52a8f6fd953e16315577c91aa163f2f8f6b48/src/JS/SAPC_0_98G_4g_minimal.js#L7

So how am I allowed to use that algorithm in my library (intended to be dual-licensed under the Apache 2.0 and MIT licenses) then? I'm not aware of how far or early WCAG 3 in development is, it'd be nice to implement now what will be effect in a few years though.

**Comment by MultisampledNight** [comment]
Oh, wow, I missed this paragraph: https://github.com/Myndex/SAPC-APCA#w3-licensed-files-moved. Then this repository just as a "non-official" branchoff makes a lot of sense.

**Comment by Myndex** [comment, marked as answer]
Hi @MultisampledNight  sorry for the delay in reply,

This is the main repository, however, the files here are not available for any form of continuous license as they are beta and in development, and due to the nature of the intended purpose (aligning with a defined standard) it is important to be able to revoke obsolete code.

There have already been issues with obsolete experimental code persisting and causing confusion, and several cases of misuse of the code and intellectual property.

The W3 version is stable, and not likely to change substantially (the possible exception is the lookup table, so the lookup table is here in the main repo instead of in the W3 repo)—it is to be licensed to the W3, and their license is similar to the MIT license.

If you do use anything in this repo, PLEASE be vigilant is maintaining and updating code when updates are posted here!


## THAT SAID:

All public licenses for APCA prohibit use in medical, clinical, aerospace, or transportation applications, and this disclaimer must accompany any use. Further, licenses shall not extend to any pending or future patents.

What is being sorted out: there is a level of capability that is intended for public free use, but there are extensions and additional functionality that may not be open source. Nevertheless it is my intention that eventually the final stable versions for general use will be.


Thank you!

Andy

**Reply by MultisampledNight** [reply]
Thank you for the thorough explanation! For the time being, I'll fallback to WCAG 2.x' contrast calculation procedure then.

**Reply by Myndex** [reply]
Hi @MultisampledNight  just be aware that the WCAG 2 contrast math is not accurate, and does not follow human perception (not even close). If you are interested in doing calculations with color, then at the very least you'd want to be using something like CIELUV or CIECAM16, etc. and never WCAG 2 contrast math, as it is functionally incorrect. You might also like HSluv.org

WCAG2 contrast math can not calculate for dark mode, is inaccurate for light text on a darker BG, and only provides a "reasonable" result for the narrow use case of a background lighter than about ` #aaaaaa `.

As I mentioned, the apca-w3 version is useable for your project, and it available via npm as ` npm i apca-w3 `
